File: python/edit_distance.py
class Solution(object):
    # O(n * m) time | O(n * m) space
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        dp = [ [ 0 ] * (len(word1) + 1) for _ in range(len(word2) + 1) ]

        for r in range(len(dp)):
            dp[r][0] = r
        for c in range(len(dp[0])):
            dp[0][c] = c

        for r in range(1, len(dp)):
            for c in range(1, len(dp[r])):
                if word2[r - 1] == word1[c - 1]:
                    dp[r][c] = dp[r - 1][c - 1]
                else:
                    dp[r][c] = 1 + min(dp[r - 1][c], dp[r][c - 1], dp[r - 1][c - 1])

        return dp[-1][-1]


solution = Solution()

word1 = "horse"
word2 = "ros"

print(solution.minDistance(word1, word2))

File: src/main/java/december_2025/EditDistance.java
package december_2025;

public class EditDistance {

    public static void main(String[] args) {
        String word1 = "horse";
        String word2 = "ros";

        EditDistance editDistance = new EditDistance();
        int result = editDistance.minDistance(word1, word2);
        System.out.println(result);
    }

    // O(n * m) time | O(n * m) space
    public int minDistance(String word1, String word2) {

        int[][] dp = new int[word2.length() + 1][word1.length() + 1];
        for (int r = 0; r < dp.length; r++) {
            dp[r][0] = r;
        }
        for (int c = 0; c < dp[0].length; c++) {
            dp[0][c] = c;
        }

        for (int r = 1; r < dp.length; r++) {
            for (int c = 1; c < dp[r].length; c++) {
                if (word2.charAt(r - 1) != word1.charAt(c - 1)) {
                    dp[r][c] = 1 + Math.min(dp[r - 1][c], Math.min(dp[r - 1][c - 1], dp[r][c - 1]));
                } else {
                    dp[r][c] = dp[r - 1][c - 1];
                }
            }
        }

        return dp[dp.length - 1][dp[0].length - 1];
    }

}
